pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt

```diff
@@ -0,0 +1,231 @@
+import org.junit.runner.RunWith
+import java.io.File
+import java.util.concurrent.TimeUnit
+
+/**
+ * Instrumented test, which will execute on an Android device.
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+@RunWith(AndroidJUnit4::class)
+class ReportingTest {
+  val context = InstrumentationRegistry.getTargetContext()
+  val instanceId = "00000000-1241-08e9-b379-377c32cd1e84"
+  val errolClient = ErrolAPI(instanceId, "http://localhost:8080")
+  val targetCtx: Context = InstrumentationRegistry.getTargetContext()
+
+  fun getStoredDeviceId(): String? {
+    val deviceStateStore = InstanceDeviceStateStore(InstrumentationRegistry.getTargetContext(), instanceId)
+    return deviceStateStore.deviceId
+  }
+
+  companion object {
+    val secretKey = "a-really-long-secret-key-that-ends-with-hunter2"
+
+    val errol = FakeErrol(8080, secretKey)
+
+    @AfterClass
+    @JvmStatic
+    fun shutdownFakeErrol() {
+      errol.stop()
+    }
+  }
+
+  @Before
+  @After
+  fun wipeLocalState() {
+    val deviceStateStore = InstanceDeviceStateStore(InstrumentationRegistry.getTargetContext(), instanceId)
+
+    await.atMost(1, TimeUnit.SECONDS) until {
+      assertTrue(deviceStateStore.clear())
+
+      deviceStateStore.interests.size == 0 && deviceStateStore.deviceId == null
+    }
+
+    File(context.filesDir, "$instanceId.jobqueue").delete()
+  }
+
+  @Before
+  @After
+  fun wipeServerReportedNotifications() {
+    errol.getInstanceEventsStorage(instanceId).clear()
+  }
+
+  private fun assertStoredDeviceIdIsNotNull() {
+    try {
+      await.atMost(DEVICE_REGISTRATION_WAIT_SECS, TimeUnit.SECONDS) untilNotNull {
+        getStoredDeviceId()
+      }
+    } catch (e: ConditionTimeoutException) {
+      // Maybe FCM is complaining in CI, so let's pretend to have a token now
+      MessagingService.onRefreshToken!!("fake-fcm-token")
+
+      await.atMost(DEVICE_REGISTRATION_WAIT_SECS, TimeUnit.SECONDS) untilNotNull {
+        getStoredDeviceId()
+      }
+    }
+  }
+
+  @Test
+  @Ignore("The job dispatcher will schedule the job but it isn't triggered within a reasonable timeframe")
+  fun aNotificationSentFromPusherShouldBeReportedTestBroken() {
+    // Start the SDK
+    val pni = PushNotificationsInstance(context, instanceId)
+    pni.start()
+
+    // A device ID should have been stored
+    assertStoredDeviceIdIsNotNull()
+    val storedDeviceId = getStoredDeviceId()
+
+    // and the device id should exist in the server
+    assertNotNull(errolClient.getDevice(storedDeviceId!!))
+
+    // send a notification
+    val i = Intent()
+    i.action = "com.google.android.c2dm.intent.RECEIVE"
+    i.`package` = targetCtx.packageName
+
+    val bundle = Bundle()
+    bundle.putString("google.delivered_priority", "high")
+    bundle.putLong("google.sent_time", 1574938426317)
+    bundle.putLong("google.ttl", 2419200)
+    bundle.putString("google.original_priority", "high")
+    bundle.putString("gcm.notification.e", "1")
```

Comment:
I don't think it relies on any undocumented internals. It's true that I've grabbed all the properties that a PN had, but it doesn't mean our reporting relies on it.

I wanted to make the test more realistic by not leaving those out (how do we test that we don't fail if they are included?). I can write a comment explaining this in the code to prevent this confusion.